### PR TITLE
Store the class object itself

### PR DESCRIPTION
**Pull request by jhaber**
Fix for #3474. This updates the `SerializedForm` to include the `Class` object rather than just the class name as suggested here https://github.com/google/protobuf/issues/3474#issuecomment-401096085 This shifts the burden of finding the right classloader from us to the JVM (which will hopefully handle it better). Based on my understanding of the Java serialization format, this change should be both forwards and backwards compatible

### CODE REVIEW

**Comment by xfxyjwf** [comment]
Thanks. This looks good to me.

@ejona86 can you also help take a look at this PR?

**Comment by jhaber** [comment]
I can manually verify forwards/backwards compatibility, and it should be easy enough to add a unit test that verifies the new code can deserialize the old serialization format (if you don't mind a hard-coded byte array in the test). Unfortunately, I can't think of a reasonable way to write a unit test that verifies the old code can deserialize the new serialization format.

**Comment by ejona86** [comment]
I don't know all the rules for forward/backward compatibility, so testing those sounds really good to me.

**Comment by jhaber** [comment]
Sure thing, I'll try to get to that tomorrow

**Comment by jhaber** [comment]
I pushed a unit test for backwards compatibility (new code can read old serialized data), and I tested forwards compatibility manually by serializing an `Int32Value` on my fork and deserializing it in a test case running against `com.google.protobuf:protobuf-java:3.5.1`:

```java
@Test
public void testForwardsCompatibility() throws Exception {
  byte[] int32ValueBytes = new byte[]{
      -84, -19, 0, 5, 115, 114, 0, 55, 99, 111, 109, 46, 103, 111, 111,
      103, 108, 101, 46, 112, 114, 111, 116, 111, 98, 117, 102, 46, 71,
      101, 110, 101, 114, 97, 116, 101, 100, 77, 101, 115, 115, 97, 103,
      101, 76, 105, 116, 101, 36, 83, 101, 114, 105, 97, 108, 105, 122,
      101, 100, 70, 111, 114, 109, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 3, 91,
      0, 7, 97, 115, 66, 121, 116, 101, 115, 116, 0, 2, 91, 66, 76, 0,
      12, 109, 101, 115, 115, 97, 103, 101, 67, 108, 97, 115, 115, 116,
      0, 17, 76, 106, 97, 118, 97, 47, 108, 97, 110, 103, 47, 67, 108,
      97, 115, 115, 59, 76, 0, 16, 109, 101, 115, 115, 97, 103, 101, 67,
      108, 97, 115, 115, 78, 97, 109, 101, 116, 0, 18, 76, 106, 97, 118,
      97, 47, 108, 97, 110, 103, 47, 83, 116, 114, 105, 110, 103, 59,
      120, 112, 117, 114, 0, 2, 91, 66, -84, -13, 23, -8, 6, 8, 84, -32,
      2, 0, 0, 120, 112, 0, 0, 0, 2, 8, 123, 118, 114, 0, 30, 99, 111,
      109, 46, 103, 111, 111, 103, 108, 101, 46, 112, 114, 111, 116, 111,
      98, 117, 102, 46, 73, 110, 116, 51, 50, 86, 97, 108, 117, 101, 0, 0,
      0, 0, 0, 0, 0, 0, 2, 0, 2, 66, 0, 21, 109, 101, 109, 111, 105, 122,
      101, 100, 73, 115, 73, 110, 105, 116, 105, 97, 108, 105, 122, 101,
      100, 73, 0, 6, 118, 97, 108, 117, 101, 95, 120, 114, 0, 38, 99, 111,
      109, 46, 103, 111, 111, 103, 108, 101, 46, 112, 114, 111, 116, 111,
      98, 117, 102, 46, 71, 101, 110, 101, 114, 97, 116, 101, 100, 77,
      101, 115, 115, 97, 103, 101, 86, 51, 0, 0, 0, 0, 0, 0, 0, 1, 2, 0,
      1, 76, 0, 13, 117, 110, 107, 110, 111, 119, 110, 70, 105, 101, 108,
      100, 115, 116, 0, 37, 76, 99, 111, 109, 47, 103, 111, 111, 103, 108,
      101, 47, 112, 114, 111, 116, 111, 98, 117, 102, 47, 85, 110, 107,
      110, 111, 119, 110, 70, 105, 101, 108, 100, 83, 101, 116, 59, 120,
      112, 116, 0, 30, 99, 111, 109, 46, 103, 111, 111, 103, 108, 101, 46,
      112, 114, 111, 116, 111, 98, 117, 102, 46, 73, 110, 116, 51, 50, 86,
      97, 108, 117, 101
  };
  ByteArrayInputStream bais = new ByteArrayInputStream(int32ValueBytes);
  ObjectInputStream in = new ObjectInputStream(bais);
  Int32Value int32Value = (Int32Value) in.readObject();
  assertThat(int32Value.getValue()).isEqualTo(123);
}
```

**Comment by jhaber** [comment]
Also worth noting that the serialized size went from 189 bytes to 413 bytes, but this overhead should be relatively static

File: java/core/src/test/java/com/google/protobuf/GeneratedMessageTest.java [hunk 1]
```diff
@@ -853,6 +854,30 @@ public void testSerializePartial() throws Exception {
     assertEquals(expected, actual);
   }
 
+  public void testSerializeBackwardsCompatibility() throws Exception {
```
Comment (ejona86) [inline java/core/src/test/java/com/google/protobuf/GeneratedMessageTest.java]: Mention somewhere what "version" of the class this is. Namely, that it was before the introduction of the `Class`.

**Comment by jhaber** [comment]
Thanks for taking a look, updated

**Comment by kevinmat** [comment]
Hi, Is this compatible with hdp 2.5 and spark version = 1.6.2.2.5.0.0-1245? Will I need to regenerate In spark-assembly jar ?

**Comment by kevinmat** [comment]
Question here is i want to make use of this feature in version 3.7 , what changes are needed in spark assembly , would i need to upgrade to a later version of Spark and HDP as well?

**Pull request merged**